control_panel_api/authentication.py

```diff
@@ -1,37 +1,69 @@
+import logging
+
+import requests
 from django.conf import settings
-import jwt
-from jwt.exceptions import InvalidTokenError
+from jose import jwt
+from jose.exceptions import JWTError
+from requests.exceptions import RequestException
 from rest_framework.authentication import (
     BaseAuthentication,
-    get_authorization_header
+    get_authorization_header,
 )
 from rest_framework.exceptions import AuthenticationFailed
 
 from control_panel_api.models import User
 
+logger = logging.getLogger(__name__)
 
-class Auth0JWTAuthentication(BaseAuthentication):
 
-    def authenticate(self, request):
+def get_jwks():
+    response = requests.get(settings.OIDC_WELL_KNOWN_URL)
```

Comment:
Again, I think may would be cleaner catch and handle the `RequestException` from inside this method where it's more clear what's going on.

This would also simplify the caller.

Reply:
I prefer the opposite - exception handling and errors should be top level to avoid having to inspect many different methods where exception handling could be. 

The top level controlling `authenticate` function knows what error bubble up, catches and handles appropriately, in one place.

Reply:
I see what you mean and it's a fair point as these methods are "external" to the `Auth0JWTAuthentication` class so raise a `AuthenticationFailed` exception may seem a bit odd.

I suppose **if** you want to make this a bit smaller there are a couple of options:
1. Extract all the logic to get/verify/decode the JWT token in a separate class which raise specific exceptions, e.g. `KidMismatchError`, etc...which then you catch in `Auth0JWTAuthentication` and handle by raising an `AuthenticationFailed` exception
2. You move these helper methods in `Auth0JWTAuthentication` so that raise  an`AuthenticationFailed` exception inside them makes sense.

Option 1 may be a bit overkill.

Obviously, feel free to play with it if you want. It seems to me that this method is basically doing 2 things:
1. Extract/verify/decode JWT token
2. Get/create the user and return it (authenticating him)

Maybe the first part could be moved in a separate method.